### PR TITLE
Being much more permissive with redir deep link URLs, adding base64 support

### PR DESCRIPTION
**Pull request by andrewconner**
@ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KifiSiteRouter.scala [hunk 1]
```diff
@@ -79,11 +80,21 @@ class KifiSiteRouter @Inject() (
...
+        Try(Json.parse(new String(CryptoSupport.fromBase64(dataStr)).trim).as[JsObject])
+      }.toOption
+    }
+
+    parseDataString(dataStr).flatMap { data =>
       val redir = deepLinkRouter.generateRedirect(data)
       redir.map(r => Ok(html.mobile.deepLinkRedirect(r, data)))
     }.getOrElse {
-      airbrake.notify(s"Could not figure out how to redirect deep-link: $dataStr")
+      println(s"[generalRedirect] Could not figure out how to redirect deep-link: $dataStr")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KifiSiteRouter.scala]: change back to airbrake once you want to merge

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/KifiSiteRouterTest.scala [hunk 2]
```diff
@@ -301,6 +301,16 @@ class KifiSiteRouterTest extends Specification with ShoeboxApplicationInjector {
...
         route(FakeRequest("GET", "/some/path").withHeaders("user-agent" -> "Mozilla/5.0 (iPhone)")) must be404
+
+
+        // Deep link v2 redirects
+        val lid = "l8rlPD6Bk7A9" // If public id config changes, this breaks. Need to hard code because the base64 encoding below.
+
+        status(route(FakeRequest("GET", s"""/redir?data=%7B"t":"lv","lid":"$lid"%7D&kma=1""")).get) === 200
+        contentAsString(route(FakeRequest("GET", s"/redir?data=%7B%22t%22%3A%22lv%22%2C%22lid%22%3A%22$lid%22%7D")).get) must contain("/abe.z1234/most-awesome-lib")
+        status(route(FakeRequest("GET", s"/redir?data=%7B%26quot;t%26quot;:%26quot;lv%26quot;,%26quot;lid%26quot;:%26quot;$lid%26quot;%7D&kma=1")).get) === 200
+        contentAsString(route(FakeRequest("GET", s"/redir?data=eyJ0IjoibHYiLCJsaWQiOiJsOHJsUEQ2Qms3QTkifQo%3D")).get) must contain("/abe.z1234/most-awesome-lib")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/KifiSiteRouterTest.scala:311]: :+1: nice tests!

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KifiSiteRouter.scala [hunk 3]
```diff
@@ -79,11 +80,21 @@ class KifiSiteRouter @Inject() (
...
-    Json.parse(dataStr).asOpt[JsObject].flatMap { data =>
+    def parseDataString(in: String): Option[JsObject] = { // We're very permissive here.
+      Try {
+        Json.parse(dataStr).as[JsObject]
+      }.orElse {
+        Try(Json.parse(dataStr.replaceAllLiterally("&quot;", "\"")).as[JsObject])
+      }.orElse {
+        Try(Json.parse(new String(CryptoSupport.fromBase64(dataStr)).trim).as[JsObject])
+      }.toOption
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KifiSiteRouter.scala:91]: Gosh Scala is pretty

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KifiSiteRouter.scala:91]: Although you probably want to do `Json.parse(in)` not `dataStr`

**Comment by ryanpbrewster** [comment]
:+1: thanks for cleaning this up

**Comment by andrewconner** [comment]
Nice catches.

**Pull request merged**